### PR TITLE
🎲 Be able to choose the NO. gpu to run

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes
This pull request includes a variety of changes to improve flexibility and performance in the codebase, particularly related to specifying the device to use for inference. The most important changes include adding support for device id in the `Device` class and adding methods to handle them, updating the `device` parameter in multiple methods to accept a string or a `Device` enum, and updating the way accelerators are set up and managed in the engine.

Main interface changes:

* <a href="diffhunk://#diff-ab44dbdadfba37e2a86acb6110ea2147c24ae3aa9ac8df7e674143c4cfac4cefR6-R61">`olive/hardware/accelerator.py`</a>: Added support for device id in the `Device` class and added methods to handle them. <a href="diffhunk://#diff-ab44dbdadfba37e2a86acb6110ea2147c24ae3aa9ac8df7e674143c4cfac4cefR6-R61">[1]</a> <a href="diffhunk://#diff-ab44dbdadfba37e2a86acb6110ea2147c24ae3aa9ac8df7e674143c4cfac4cefR75-R82">[2]</a>
* <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L133-R133">`olive/model/__init__.py`</a>: Updated the `device` parameter in several methods to accept a string or a `Device` enum, and converted the parameter to a `Device` object if it is a string. <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L133-R133">[1]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L229-R229">[2]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L325-R329">[3]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L408-R414">[4]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L619-R625">[5]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L834-R841">[6]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L886-R906">[7]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L937-R955">[8]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L957-R965">[9]</a> <a href="diffhunk://#diff-10a7e98b2eb102b4246031efabef4995ea5c165b4b5b5c14639d51a4df349802L1044-R1052">[10]</a>
* <a href="diffhunk://#diff-f1038110c73dab1688ffe460efc340cdf2537ad870d6ef1e767c57836f73dcceL160-R164">`olive/engine/engine.py`</a>: Updated the way accelerators are set up and managed in the engine, including changes to use `device.name` instead of `device` and updating the `get_managed_supported_execution_providers` method to accept a string instead of a `Device` object. <a href="diffhunk://#diff-f1038110c73dab1688ffe460efc340cdf2537ad870d6ef1e767c57836f73dcceL160-R164">[1]</a> <a href="diffhunk://#diff-f1038110c73dab1688ffe460efc340cdf2537ad870d6ef1e767c57836f73dcceL150-R150">[2]</a>

Other important changes:

* `olive/evaluator/olive_evaluator.py`: Updated the `device` parameter in multiple methods to accept a string or a `Device` enum, allowing for more flexibility in specifying the device to use for inference. (F55dafa6)
* <a href="diffhunk://#diff-3c4b1c200e2ac992cd17d17e72a48402d7376e0dcc11d3c4a3728d630ec48097L57-R57">`test/unit_test/passes/onnx/test_perf_tuning.py`</a>: Updated performance tuning in ONNX tests, such as adding a new accelerator and changing the device value from "gpu" to "cuda". <a href="diffhunk://#diff-3c4b1c200e2ac992cd17d17e72a48402d7376e0dcc11d3c4a3728d630ec48097L57-R57">[1]</a> <a href="diffhunk://#diff-3c4b1c200e2ac992cd17d17e72a48402d7376e0dcc11d3c4a3728d630ec48097L39-R39">[2]</a>
* <a href="diffhunk://#diff-13874afac6177702a22337e2731db6da5c680ef7b1326fac40c92b955b66228dL16-R16">`olive/passes/onnx/conversion.py`</a>: Removed the `Device` import and replaced it with the `to_torch_device` method of the `AcceleratorType` class. <a href="diffhunk://#diff-13874afac6177702a22337e2731db6da5c680ef7b1326fac40c92b955b66228dL16-R16">[1]</a> <a href="diffhunk://#diff-13874afac6177702a22337e2731db6da5c680ef7b1326fac40c92b955b66228dL214-R214">[2]</a>
* <a href="diffhunk://#diff-29574b795eaec4f329266c39ced57cfcddde5dcea7e73ff66297fed88b5d0d62L72-R72">`olive/systems/python_environment/inference_runner.py`</a>: Updated the device binding for CUDA to use "cuda" instead of "gpu" if `args.device` is "cuda".
* <a href="diffhunk://#diff-2b394e839ea5381905d411fdd72a337faa167c2bfada8966d206c66e569bfa5fL97-R105">`test/unit_test/systems/test_local.py`</a>: Updated the `device` parameter to accept a string instead of a `Device` enum, allowing for more flexibility in specifying the device to use for inference.
## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/hardware/accelerator.py [hunk 1]
```diff
@@ -4,20 +4,57 @@
...
     GPU = "gpu"
     NPU = "npu"
     INTEL_MYRIAD = "intel_myriad"
+    # SNPE
+    DSP = "dsp"
+    AIP = "aip"
+
+    def __init__(self, name: str = CPU, idx: int = 0):
+        # if name is like: name:0, name:1, name:2, ...
+        name_list = name.lower().split(":")
```
[the remark this answers is not in the document]
Comment (guotuofeng) [reply olive/hardware/accelerator.py:23]: Basically, we need not change the device enum class as it is behavior like a shorthand for string representation.

Instead, we might not change so many files. from my point of view, we can pass the "cuda:0" str as string to postpone the process until it need handle torch.device. That's to say only do the necessary processing when it has to. In this way, I suppose only several files need to be changed.

[the remark this answers is not in the document]
Comment (guotuofeng) [reply olive/hardware/accelerator.py:23]: as @devang-ml said, AcceleratorSpec is the right place to handle the index

[the remark this answers is not in the document]
Comment (trajepl) [reply olive/hardware/accelerator.py]: We found a better way to restrict the cuda device. Currently it worked for torch and onnxurntime. 
export CUDA_VISIBLE_DEVICES=2

I am not sure about other npu.  
But currently, it met my requirements. I will close this PR and update our docs to tell the way to choose device.

File: olive/evaluator/olive_evaluator.py [hunk 2]
```diff
@@ -80,6 +80,7 @@ def _inference(
         dataloader: Dataset,
         post_func=None,
         device: Device = Device.CPU,
+        device_id: int = 0,
```
Comment (guotuofeng) [inline olive/evaluator/olive_evaluator.py:83]: Optional[int] = None

File: olive/systems/local.py [hunk 3]
```diff
@@ -44,7 +44,14 @@ def evaluate_model(
 
         model = model_config.create_model()
         evaluator: OliveEvaluator = OliveEvaluatorFactory.create_evaluator_for_model(model)
-        return evaluator.evaluate(model, data_root, metrics, device=device, execution_providers=execution_providers)
+        return evaluator.evaluate(
+            model,
+            data_root,
+            metrics,
+            device=device,
+            device_id=accelerator.device_id,
```
Comment (guotuofeng) [inline olive/systems/local.py:52]: azureml and docker system need handle it as well.

Comment (trajepl) [reply olive/systems/local.py:52]: azureml do not need this as it run like local system.

Docker needs this one.

**Comment by trajepl** [comment]
Do not delete this branch in case for the non-cuda device, we still need to choose the device id.

File: olive/systems/local.py (hunk 3 above)
Comment (guotuofeng) [reply olive/systems/local.py:52]: azureml does have logic to handle accelerator stuff in the entry

Comment (trajepl) [reply olive/systems/local.py:52]: Yes, somehow, we may need to re-think the part of perf-tuning to let is tune only in one EPs like other passes in the future. Tuning multi-eps were done be Lei before. 

In this way, we can only pass accelerator to evaluator.

**Pull request closed without merging**